### DISCUSSION TITLE
Deserialization issue on Resume with System.Activities.Runtime.ExecutionPropertyManager

### DISCUSSION THREAD

**Discussion by fgollasmmi** [UiPath/corewf#209] [Q&A] [answered]
We try to resume a WorkflowApplication instance from the FileInstanceStore of one of your examples and run into a Newtonsoft.Json.JsonSerializationException (see below).

The deserialization fails becaus there is no default constructor in the class System.Activities.Runtime.ExecutionPropertyManager.

However the [ExecutionPropertyManager](https://github.com/UiPath/CoreWF/blob/develop/src/UiPath.Workflow.Runtime/Runtime/ExecutionPropertyManager.cs) is flagged as [DataContract]. 

In my understanding it would be generally recommended to have a default constructor for each DataContract.

Adding a default constructor solves the issue for me.
[ae2ff8a9-2258-4d80-b1e4-e6086fd5e5d0-InstanceMetadata.txt](https://github.com/UiPath/CoreWF/files/8927110/ae2ff8a9-2258-4d80-b1e4-e6086fd5e5d0-InstanceMetadata.txt)
[ae2ff8a9-2258-4d80-b1e4-e6086fd5e5d0-InstanceData.txt](https://github.com/UiPath/CoreWF/files/8927111/ae2ff8a9-2258-4d80-b1e4-e6086fd5e5d0-InstanceData.txt)

Finally my question, whether there is a reason for not having a default constructor on System.Activities.Runtime.ExecutionPropertyManager.

The exception message and stacktrace I get on resuming a WorkflowApplication instance from an existing GUID.
```
Message: 
Newtonsoft.Json.JsonSerializationException : Unable to find a constructor to use for type System.Activities.Runtime.ExecutionPropertyManager. A class should either have a default constructor, one constructor with arguments or a constructor marked with the JsonConstructor attribute. Path '['{urn:schemas-microsoft-com:System.Activities/4.0/properties}Workflow'].Value.bookmarkMgr.bookmarks.1.ActivityInstance.propertyManager.properties', line 147, position 29.

Stack Trace: 
AsyncResult.End[TAsyncResult](IAsyncResult result) line 332
ExecuteAsyncResult.End(IAsyncResult result) line 1039
InstancePersistenceContext.OuterExecute(InstanceHandle initialInstanceHandle, InstancePersistenceCommand command, TimeSpan timeout) line 757
InstanceStore.Execute(InstanceHandle handle, InstancePersistenceCommand command, TimeSpan timeout) line 79
PersistenceManager.Load(TimeSpan timeout) line 412
WorkflowApplication.LoadValues(PersistenceManager persistenceManager, TimeoutHelper timeoutHelper, Boolean loadAny) line 1929
WorkflowApplication.LoadCore(TimeoutHelper timeoutHelper, Boolean loadAny, IDictionary`2 values) line 1765
WorkflowApplication.Load(Guid instanceId, TimeSpan timeout) line 1726
WorkflowApplication.Load(Guid instanceId) line 1684
```

-- EDIT --
Attached both, the InstanceData and InstanceMetadata file from FileInstanceStore.
-- END EDIT --

**Comment by fanzh0513** [comment]
Dear

I have the same question, but I think that is not CoreWF's problem.  you can assign a JsonConverter for the ExecutionPropertyManager class. 

the sample code is :

        // the first, you must assign a JsonConverter.
        private readonly JsonSerializerSettings _jsonSerializerSettings = new JsonSerializerSettings
        {
            TypeNameHandling = TypeNameHandling.Auto,
            TypeNameAssemblyFormatHandling = TypeNameAssemblyFormatHandling.Simple,
            ConstructorHandling = ConstructorHandling.AllowNonPublicDefaultConstructor,
            ObjectCreationHandling = ObjectCreationHandling.Replace,
            PreserveReferencesHandling = PreserveReferencesHandling.Objects,
            Converters = new[]
            {
                new ExecutionPropertyManagerConverter()
            }
        };


 class ExecutionPropertyManagerConverter : JsonConverter
        {
            public override bool CanConvert(Type objectType)
            {
                if (objectType.Name == "ExecutionPropertyManager")
                    return true;

                return false;
            }

            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
            {
                object dictExecutionProperty = default;
                int exclusiveHandleCount = default;

                PropertyInfo[] properties = objectType.GetProperties(BindingFlags.NonPublic | BindingFlags.Instance).Where(i => i.GetCustomAttribute<DataMemberAttribute>(false) != null).ToArray();
                if (properties.Length > 0)
                {
                    foreach (KeyValuePair<string, JToken> item in serializer.Deserialize<JObject>(reader))
                    {
                        PropertyInfo property = properties.FirstOrDefault(i => i.GetCustomAttribute<DataMemberAttribute>().Name == item.Key);
                        switch (item.Key)
                        {
                            case "properties":
                                dictExecutionProperty = JsonConvert.DeserializeObject(item.Value?.ToString(), property.PropertyType);
                                break;
                            case "exclusiveHandleCount":
                                exclusiveHandleCount = (int)item.Value;
                                break;
                            default:
                                throw new NotSupportedException(item.Key);
                        }
                    }
                }

                object o = Activator.CreateInstance(objectType, BindingFlags.Instance | BindingFlags.NonPublic, null, new object[] { null, dictExecutionProperty }, CultureInfo.CurrentCulture);

                // process exclusiveHandleCount value.
                PropertyInfo p = properties.FirstOrDefault(i => i.GetCustomAttribute<DataMemberAttribute>().Name == "exclusiveHandleCount");
                DataMemberAttribute attribute = p.GetCustomAttribute<DataMemberAttribute>();

                if(attribute.EmitDefaultValue && exclusiveHandleCount == default)
                {
                    p.SetValue(o, exclusiveHandleCount);
                }

                return o;
            }

            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
            {
                PropertyInfo[] properties = value.GetType().GetProperties(BindingFlags.NonPublic | BindingFlags.Instance).Where(i => i.GetCustomAttribute<DataMemberAttribute>(false) != null).ToArray();
                if (properties.Length > 0)
                {
                    writer.WriteStartObject();
                    foreach (PropertyInfo property in properties)
                    {
                        DataMemberAttribute attribute = property.GetCustomAttribute<DataMemberAttribute>();

                        object o = property.GetValue(value);
                        if (o != default)
                        {
                            writer.WritePropertyName(attribute.Name);
                            switch (attribute.Name)
                            {
                                case "properties":
                                    writer.WriteRawValue(JsonConvert.SerializeObject(o, Formatting.Indented));
                                    break;
                                case "exclusiveHandleCount":
                                    writer.WriteValue(o);
                                    break;
                                default:
                                    throw new NotSupportedException(attribute.Name);
                            }
                        }
                        else if (attribute.EmitDefaultValue)
                        {
                            writer.WritePropertyName(attribute.Name);
                            writer.WriteNull();
                        }
                    }

                    writer.WriteEndObject();
                }
            }
        }

**Comment by lbargaoanu** [comment, marked as answer]
https://uipath.visualstudio.com/Public.Feeds/_artifacts/feed/UiPath-Internal/NuGet/UiPath.Workflow/overview/6.0.0-20221018-03
